tephra-core/src/main/java/co/cask/tephra/persist/HDFSTransactionLogReaderV2.java

```diff
@@ -0,0 +1,116 @@
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * {@link TransactionLogReader} that can read v2 version of Transaction logs. The logs are expected to
+ * have a sequence of {@link TransactionEdit}s along with commit markers that indicates the size of the batch of
+ * {@link TransactionEdit}s that were synced together. If the expected number of {@link TransactionEdit}s are not
+ * present then that set of {@link TransactionEdit}s are discarded.
+ */
+public class HDFSTransactionLogReaderV2 extends AbstractHDFSLogReader {
+  private static final Logger LOG = LoggerFactory.getLogger(HDFSTransactionLogReaderV2.class);
+
+  private final Queue<TransactionEdit> transactionEdits;
+  private final LongWritable key;
+
+  public HDFSTransactionLogReaderV2(SequenceFile.Reader reader) {
+    super(reader);
+    this.transactionEdits = new ArrayDeque<>();
+    this.key = new LongWritable();
+  }
+
+  @Override
+  public TransactionEdit next() throws IOException {
+    return next(null);
+  }
+
+  @Override
+  public TransactionEdit next(TransactionEdit reuse) throws IOException {
+    if (isClosed()) {
+      return null;
+    }
+
+    if (transactionEdits.size() != 0) {
+      return transactionEdits.remove();
+    }
+
+    // Read edits until we reach the commit marker
+    populateTransactionEdits();
+    return transactionEdits.poll();
+  }
+
+  private void populateTransactionEdits() throws IOException {
+    // read the marker to determine numEntries to read.
+    int numEntries = fetchMarker();
+    for (int i = 0; i < numEntries; i++) {
+      TransactionEdit edit = new TransactionEdit();
+      try {
+        if (reader.next(key, edit)) {
+          transactionEdits.add(edit);
+        } else {
+          throw new EOFException("Attempt to read TransactionEdit failed.");
+        }
+      } catch (EOFException e) {
+        // we have reached EOF before reading back numEntries, we clear the partial list and return.
+        LOG.warn("Reached EOF in log before reading {} entries. Ignoring all {} edits since the last marker",
+                 numEntries, transactionEdits.size(), e);
+        transactionEdits.clear();
+      }
+    }
+  }
+
+  private int fetchMarker() throws IOException {
+    byte[] keyBytes = TxConstants.TransactionLog.NUM_ENTRIES_APPENDED.getBytes();
+    try (DataOutputBuffer rawKey = new DataOutputBuffer();
+         ByteArrayOutputStream value = new ByteArrayOutputStream(Integer.SIZE);
+         DataOutputStream outputStream = new DataOutputStream(value)) {
+      SequenceFile.ValueBytes valueBytes = reader.createValueBytes();
+      int status = reader.nextRaw(rawKey, valueBytes);
+
+      //Check if the marker key is valid
+      if (status != -1 && rawKey.getLength() == keyBytes.length && Bytes.indexOf(rawKey.getData(), keyBytes) != -1) {
```

Comment:
It would be more readable if `rawKey.getLength() == keyBytes.length && Bytes.indexOf(rawKey.getData(), keyBytes) != -1` is extracted into a private method, right?